core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
     }
   }
 
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    numSentBytesQueued += numBytes;
+  }
+
+  /**
+   * Event handler to be called by the subclass when a number of bytes has been sent to the remote
+   * endpoint. This method will invoke {@link StreamListener#onReady(int)} when appropriate.
+   *
+   * @param numBytes the number of bytes that were sent.
+   */
+  protected final void onSentBytes(int numBytes) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
+      // TODO(nathanmittler): Consider selecting a better value for numMessages.
+      listener().onReady();
```

Comment:
If CLOSED, then we must not call `onReady()`.

On client-side, it isn't strictly necessary to call `onReady()` if HALF_CLOSED, but that seems like it would be a bit of a pain, since this is shared between client and server. I'm okay if we don't do this for now.